File: user_app/src/services/credentialStorage.ts
/**
 * Credential-storage abstraction for device credentials.
 *
 * The plaintext API key is returned **only** at issuance or rotation.
 * This abstraction ensures the same higher-level workflow works across
 * development simulations and production platforms.
 *
 * Implementations:
 * - `SimulationStorage`  — sessionStorage/localStorage (for dev/testing)
 * - `LinuxFileStorage`   — file on disk with strict permissions (0600)
 * - `WindowsCredManager` — placeholder for Windows Credential Manager / DPAPI
 * - `AndroidKeystore`    — placeholder for Android Keystore
 */

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface DeviceCredentials {
  deviceId: string
  apiKey: string
  siteId?: string
  deviceName?: string
  deviceType?: string
  deviceOs?: string
  enrollmentMethod?: string
}

// ---------------------------------------------------------------------------
// Abstract interface
// ---------------------------------------------------------------------------

export interface CredentialStorage {
  /** Persist credentials after a successful provision or claim. */
  save(creds: DeviceCredentials): Promise<void>

  /** Retrieve the stored API key.  Returns ``null`` if not provisioned. */
  getApiKey(): Promise<string | null>

  /** Retrieve the stored device ID.  Returns ``null`` if not provisioned. */
  getDeviceId(): Promise<string | null>

  /** Retrieve a metadata field by key. */
  getMetadata(key: string): Promise<string | null>

  /** Remove all stored credentials (unpair / factory-reset). */
  clear(): Promise<void>

  /** ``true`` if credentials are present (device is provisioned). */
  isProvisioned(): Promise<boolean>
}

// ---------------------------------------------------------------------------
// Simulation storage (sessionStorage + localStorage)
// ---------------------------------------------------------------------------

const LS_PREFIX = 'homepot_'
const SS_API_KEY = 'homepot_api_key'

function lsKey(key: string): string {
  return `${LS_PREFIX}${key}`
}

export class SimulationStorage implements CredentialStorage {
  async save(creds: DeviceCredentials): Promise<void> {
    localStorage.setItem(lsKey('device_id'), creds.deviceId)
    localStorage.setItem(lsKey('device_name'), creds.deviceName ?? '')
    localStorage.setItem(lsKey('device_type'), creds.deviceType ?? '')
    localStorage.setItem(lsKey('device_os'), creds.deviceOs ?? '')
    localStorage.setItem(lsKey('enrollment_method'), creds.enrollmentMethod ?? 'self-enrolled')
    if (creds.siteId) {
      localStorage.setItem(lsKey('site_id'), creds.siteId)
    }
    // API key goes to sessionStorage (cleared when tab closes)
    sessionStorage.setItem(SS_API_KEY, creds.apiKey)
  }

  async getApiKey(): Promise<string | null> {
    return sessionStorage.getItem(SS_API_KEY)
  }

  async getDeviceId(): Promise<string | null> {
    return localStorage.getItem(lsKey('device_id'))
  }

  async getMetadata(key: string): Promise<string | null> {
    return localStorage.getItem(lsKey(key))
  }

  async clear(): Promise<void> {
    const keysToRemove: string[] = []
    for (let i = 0; i < localStorage.length; i++) {
      const k = localStorage.key(i)
      if (k?.startsWith(LS_PREFIX)) keysToRemove.push(k)
    }
    keysToRemove.forEach((k) => localStorage.removeItem(k))
    sessionStorage.removeItem(SS_API_KEY)
  }

  async isProvisioned(): Promise<boolean> {
    return localStorage.getItem(lsKey('device_id')) !== null
  }
}

// ---------------------------------------------------------------------------
// Linux file storage (file on disk with strict permissions)
// ---------------------------------------------------------------------------

/**
 * LinuxFileStorage stores credentials in a JSON file at ``~/.homepot/credentials``
 * with ``0600`` permissions.  Falls back to ``SimulationStorage`` if the
 * filesystem API is unavailable (browser environment).
 */
export class LinuxFileStorage implements CredentialStorage {
  private fallback: SimulationStorage
  private filePath = ''

  constructor() {
    this.fallback = new SimulationStorage()
    // Attempt to resolve a writable path (only in Node/Electron/Tauri)
    try {
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      const os = require('os') as { homedir(): string }
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      const path = require('path') as { join(...parts: string[]): string }
      this.filePath = path.join(os.homedir(), '.homepot', 'credentials')
    } catch {
      // Browser environment — use simulation storage
    }
  }

  private async readFile(): Promise<Record<string, string> | null> {
    if (!this.filePath) return null
    try {
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      const fs = require('fs') as {
        existsSync(p: string): boolean
        readFileSync(p: string, enc: string): string
        mkdirSync(p: string, opts: { recursive: boolean }): void
        writeFileSync(p: string, data: string, opts: { mode: number }): void
        chmodSync(p: string, mode: number): void
      }
      if (!fs.existsSync(this.filePath)) return null
      const raw = fs.readFileSync(this.filePath, 'utf-8')
      return JSON.parse(raw) as Record<string, string>
    } catch {
      return null
    }
  }

  private async writeFile(data: Record<string, string>): Promise<void> {
    if (!this.filePath) return
    try {
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      const fs = require('fs') as {
        existsSync(p: string): boolean
        mkdirSync(p: string, opts: { recursive: boolean }): void
        writeFileSync(p: string, data: string, opts: { mode: number }): void
        chmodSync(p: string, mode: number): void
      }
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      const path = require('path') as { dirname(p: string): string }
      const dir = path.dirname(this.filePath)
      if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true })
      }
      fs.writeFileSync(this.filePath, JSON.stringify(data, null, 2), { mode: 0o600 })
      fs.chmodSync(this.filePath, 0o600)
    } catch {
      // Fallback to simulation storage
      await this.fallback.clear()
    }
  }

  async save(creds: DeviceCredentials): Promise<void> {
    if (this.filePath) {
      const data: Record<string, string> = {
        device_id: creds.deviceId,
        api_key: creds.apiKey,
      }
      if (creds.siteId) data.site_id = creds.siteId
      if (creds.deviceName) data.device_name = creds.deviceName
      if (creds.deviceType) data.device_type = creds.deviceType
      if (creds.deviceOs) data.device_os = creds.deviceOs
      if (creds.enrollmentMethod) data.enrollment_method = creds.enrollmentMethod
      await this.writeFile(data)
    }
    // Also save to simulation storage for backward compat
    await this.fallback.save(creds)
  }

  async getApiKey(): Promise<string | null> {
    if (this.filePath) {
      const data = await this.readFile()
      if (data?.api_key) return data.api_key
    }
    return this.fallback.getApiKey()
  }

  async getDeviceId(): Promise<string | null> {
    if (this.filePath) {
      const data = await this.readFile()
      if (data?.device_id) return data.device_id
    }
    return this.fallback.getDeviceId()
  }

  async getMetadata(key: string): Promise<string | null> {
    if (this.filePath) {
      const data = await this.readFile()
      if (data?.[key]) return data[key]
    }
    return this.fallback.getMetadata(key)
  }

  async clear(): Promise<void> {
    if (this.filePath) {
      try {
        // eslint-disable-next-line @typescript-eslint/no-require-imports
        const fs = require('fs') as { unlinkSync(p: string): void }
        if (fs.existsSync(this.filePath)) {
          fs.unlinkSync(this.filePath)
        }
      } catch {
        // ignore
      }
    }
    await this.fallback.clear()
  }

  async isProvisioned(): Promise<boolean> {
    if (this.filePath) {
      const data = await this.readFile()
      if (data?.device_id) return true
    }
    return this.fallback.isProvisioned()
  }
}

// ---------------------------------------------------------------------------
// Platform-aware factory
// ---------------------------------------------------------------------------

/**
 * Returns the appropriate ``CredentialStorage`` implementation for the
 * current platform.
 *
 * - **Linux** (Node/Electron/Tauri) → ``LinuxFileStorage``
 * - **Browser** → ``SimulationStorage``
 * - **Windows** → ``WindowsCredManager`` *(placeholder, falls back to simulation)*
 * - **Android** → ``AndroidKeystore`` *(placeholder, falls back to simulation)*
 */
export function createCredentialStorage(): CredentialStorage {
  // In a browser context (no `process` or `require`), use simulation
  if (typeof window !== 'undefined' && typeof document !== 'undefined') {
    // Check for Android bridge
    if (navigator.userAgent?.toLowerCase().includes('android')) {
      // AndroidKeystore placeholder — uses simulation for now
      console.warn('[credentialStorage] Android Keystore not yet implemented; using SimulationStorage')
      return new SimulationStorage()
    }
    return new SimulationStorage()
  }

  // Server-side / Electron / Tauri — detect platform
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const os = require('os') as { platform(): string }
    const platform = os.platform()
    if (platform === 'linux') {
      return new LinuxFileStorage()
    }
    if (platform === 'win32') {
      console.warn('[credentialStorage] Windows Credential Manager not yet implemented; using LinuxFileStorage')
      return new LinuxFileStorage()
    }
  } catch {
    // Fallback
  }

  return new SimulationStorage()
}

// ---------------------------------------------------------------------------
// Singleton instance
// ---------------------------------------------------------------------------

/** Application-wide credential storage instance. */
export const credentialStorage: CredentialStorage = createCredentialStorage()

File: backend/src/homepot/agent/credential_storage.py
"""Credential-storage abstraction for device credentials.

Mirrors the TypeScript CredentialStorage interface from
user_app/src/services/credentialStorage.ts.

The plaintext API key is returned **only** at issuance or rotation.
This abstraction ensures the same higher-level workflow works across
development simulations and production platforms.

Implementations:
- ``SimulationStorage`` — in-memory dict / temp file (for dev/testing)
- ``LinuxFileStorage`` — file on disk with strict permissions (0o600)
- ``KeyringCredentialStorage`` — OS keyring via the ``keyring`` library
- ``WindowsCredManager`` — Windows Credential Manager via ``keyring``
- ``WindowsFileStorage`` — file fallback on Windows (``%PROGRAMDATA%``)
- ``AndroidKeystore`` — placeholder for Android Keystore
"""

from abc import ABC, abstractmethod
import json
import logging
import os
from pathlib import Path
import sys
from typing import Dict, Optional

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Types
# ---------------------------------------------------------------------------

DeviceCredentials = Dict[str, str]


# ---------------------------------------------------------------------------
# Abstract interface
# ---------------------------------------------------------------------------


class CredentialStorage(ABC):
    """Abstract interface for persisting device credentials."""

    @abstractmethod
    def save(self, creds: DeviceCredentials) -> None:
        """Persist credentials after a successful provision or claim."""

    @abstractmethod
    def get_api_key(self) -> Optional[str]:
        """Retrieve the stored API key.  Returns ``None`` if not provisioned."""

    @abstractmethod
    def get_device_id(self) -> Optional[str]:
        """Retrieve the stored device ID.  Returns ``None`` if not provisioned."""

    @abstractmethod
    def get_metadata(self, key: str) -> Optional[str]:
        """Retrieve a metadata field by key."""

    @abstractmethod
    def clear(self) -> None:
        """Remove all stored credentials (unpair / factory-reset)."""

    @abstractmethod
    def is_provisioned(self) -> bool:
        """Return True if credentials are present."""

    # ------------------------------------------------------------------
    # Concrete convenience helpers (reuse the abstract interface above)
    # ------------------------------------------------------------------

    def get_backend_url(self) -> Optional[str]:
        """Return the configured backend URL or None."""
        return self.get_metadata("backend_url")

    def set_backend_url(self, url: str) -> None:
        """Persist the backend URL so the agent knows where to connect."""
        self.save({"backend_url": url})

    def get_tls_verify(self) -> bool:
        """Return whether TLS certificate verification is enabled.

        Defaults to ``True``.  Once explicitly set to ``"false"`` via
        ``set_tls_config`` it returns ``False``.
        """
        val = self.get_metadata("tls_verify")
        return val != "false" if val else True

    def get_tls_ca_cert(self) -> Optional[str]:
        """Return the path (or PEM content) of the TLS CA certificate."""
        return self.get_metadata("tls_ca_cert")

    def get_tls_client_cert(self) -> Optional[str]:
        """Return the path (or PEM content) of the TLS client cert."""
        return self.get_metadata("tls_client_cert")

    def get_tls_client_key(self) -> Optional[str]:
        """Return the path (or PEM content) of the TLS client key."""
        return self.get_metadata("tls_client_key")

    def set_tls_config(
        self,
        verify: bool = True,
        ca_cert: Optional[str] = None,
        client_cert: Optional[str] = None,
        client_key: Optional[str] = None,
    ) -> None:
        """Persist TLS configuration alongside credentials.

        Parameters
        ----------
        verify:
            Whether to verify the server's TLS certificate.
        ca_cert:
            Path to a CA certificate file or PEM-encoded content.
        client_cert:
            Path to a client certificate file or PEM-encoded content
            (mutual TLS).
        client_key:
            Path to the client private key file or PEM-encoded content.
        """
        data: Dict[str, str] = {"tls_verify": str(verify).lower()}
        if ca_cert is not None:
            data["tls_ca_cert"] = ca_cert
        if client_cert is not None:
            data["tls_client_cert"] = client_cert
        if client_key is not None:
            data["tls_client_key"] = client_key
        self.save(data)


# ---------------------------------------------------------------------------
# Simulation storage (in-memory)
# ---------------------------------------------------------------------------


class SimulationStorage(CredentialStorage):
    """In-memory credential storage for testing and development."""

    def __init__(self) -> None:
        """Initialize an empty in-memory credential store."""
        self._data: Dict[str, str] = {}

    def save(self, creds: DeviceCredentials) -> None:
        """Save credentials to the in-memory store."""
        self._data.update(creds)

    def get_api_key(self) -> Optional[str]:
        """Return the stored API key or None."""
        return self._data.get("api_key")

    def get_device_id(self) -> Optional[str]:
        """Return the stored device ID or None."""
        return self._data.get("device_id")

    def get_metadata(self, key: str) -> Optional[str]:
        """Return a metadata field by key or None."""
        return self._data.get(key)

    def clear(self) -> None:
        """Remove all stored credentials."""
        self._data.clear()

    def is_provisioned(self) -> bool:
        """Return True if device_id is present in the store."""
        return "device_id" in self._data


# ---------------------------------------------------------------------------
# Linux file storage (file on disk with strict permissions)
# ---------------------------------------------------------------------------


class LinuxFileStorage(CredentialStorage):
    """Stores credentials in a JSON file with ``0o600`` permissions.

    Default location: ``~/.homepot/credentials``
    """

    def __init__(self, file_path: Optional[Path] = None) -> None:
        """Initialize Linux file storage at the given path."""
        self._file_path = file_path or Path.home() / ".homepot" / "credentials"

    def _read(self) -> Dict[str, str]:
        """Read credentials from the JSON file. Returns empty dict on error."""
        if not self._file_path.exists():
            return {}
        try:
            raw = self._file_path.read_text("utf-8")
            return dict(json.loads(raw))
        except (json.JSONDecodeError, OSError) as exc:
            logger.warning("Failed to read credentials file: %s", exc)
            return {}

    def _write(self, data: Dict[str, str]) -> None:
        """Write credentials to the JSON file with 0o600 permissions."""
        self._file_path.parent.mkdir(parents=True, exist_ok=True)
        self._file_path.write_text(
            json.dumps(data, indent=2),
            encoding="utf-8",
        )
        self._file_path.chmod(0o600)

    def save(self, creds: DeviceCredentials) -> None:
        """Save credentials to the JSON file."""
        data = self._read()
        data.update(creds)
        self._write(data)

    def get_api_key(self) -> Optional[str]:
        """Return the stored API key or None."""
        return self._read().get("api_key")

    def get_device_id(self) -> Optional[str]:
        """Return the stored device ID or None."""
        return self._read().get("device_id")

    def get_metadata(self, key: str) -> Optional[str]:
        """Return a metadata field by key or None."""
        return self._read().get(key)

    def clear(self) -> None:
        """Remove the credentials file."""
        if self._file_path.exists():
            self._file_path.unlink()

    def is_provisioned(self) -> bool:
        """Return True if the file contains a device_id."""
        return self._file_path.exists() and "device_id" in self._read()


# ---------------------------------------------------------------------------
# Keyring storage (OS keyring via the ``keyring`` library)
# ---------------------------------------------------------------------------


class KeyringCredentialStorage(CredentialStorage):
    """Stores credentials in the OS keyring via the ``keyring`` library.

    The entire credentials dict is serialised as JSON and stored under
    *service* ``"homepot-agent"`` / *username* ``"credentials"``.

    Requires the ``keyring`` package (install with ``pip install homepot-client[agent]``).
    When the keyring backend is unavailable (e.g. headless server without a
    keyring daemon) the factory falls back to ``LinuxFileStorage``.
    """

    _SERVICE = "homepot-agent"
    _USERNAME = "credentials"

    def __init__(self) -> None:
        """Initialize keyring storage (lazy import of ``keyring``)."""
        try:
            import keyring as _kr

            self._kr = _kr
        except ImportError:
            raise ImportError(
                "keyring package is required. Install with: pip install homepot-client[agent]"
            )

    def _read(self) -> Dict[str, str]:
        """Retrieve credentials from the OS keyring."""
        try:
            raw = self._kr.get_password(self._SERVICE, self._USERNAME)
            if raw is None:
                return {}
            return dict(json.loads(raw))
        except Exception as exc:
            logger.warning("Failed to read credentials from keyring: %s", exc)
            return {}

    def _write(self, data: Dict[str, str]) -> None:
        """Store credentials in the OS keyring."""
        try:
            self._kr.set_password(
                self._SERVICE,
                self._USERNAME,
                json.dumps(data, indent=2),
            )
        except Exception as exc:
            logger.error("Failed to write credentials to keyring: %s", exc)
            raise

    def save(self, creds: DeviceCredentials) -> None:
        """Merge and persist credentials in the OS keyring."""
        data = self._read()
        data.update(creds)
        self._write(data)

    def get_api_key(self) -> Optional[str]:
        """Return the stored API key or None."""
        return self._read().get("api_key")

    def get_device_id(self) -> Optional[str]:
        """Return the stored device ID or None."""
        return self._read().get("device_id")

    def get_metadata(self, key: str) -> Optional[str]:
        """Return a metadata field by key or None."""
        return self._read().get(key)

    def clear(self) -> None:
        """Remove all stored credentials from the OS keyring."""
        try:
            self._kr.delete_password(self._SERVICE, self._USERNAME)
        except self._kr.errors.PasswordDeleteError:
            pass
        except Exception as exc:
            logger.warning("Failed to clear credentials from keyring: %s", exc)

    def is_provisioned(self) -> bool:
        """Return True if the keyring contains a device_id."""
        return self._read().get("device_id") is not None


# ---------------------------------------------------------------------------
# Windows Credential Manager (via the ``keyring`` library)
# ---------------------------------------------------------------------------


class WindowsCredManager(CredentialStorage):
    """Stores credentials in Windows Credential Manager via the ``keyring`` library.

    The entire credentials dict is serialised as JSON and stored under
    *service* ``"homepot-agent-windows"`` / *username* ``"credentials"``.

    Requires the ``keyring`` package (install with ``pip install homepot-client[agent]``).
    When the keyring backend is unavailable (e.g. headless or CI without
    Credential Manager service) the factory falls back to ``WindowsFileStorage``.
    """

    _SERVICE = "homepot-agent-windows"
    _USERNAME = "credentials"

    def __init__(self) -> None:
        """Initialize Windows Credential Manager storage."""
        try:
            import keyring as _kr

            self._kr = _kr
        except ImportError:
            raise ImportError(
                "keyring package is required. Install with: pip install homepot-client[agent]"
            )

    def _read(self) -> Dict[str, str]:
        """Retrieve credentials from Windows Credential Manager."""
        try:
            raw = self._kr.get_password(self._SERVICE, self._USERNAME)
            if raw is None:
                return {}
            return dict(json.loads(raw))
        except Exception as exc:
            logger.warning("Failed to read credentials from keyring: %s", exc)
            return {}

    def _write(self, data: Dict[str, str]) -> None:
        """Store credentials in Windows Credential Manager."""
        try:
            self._kr.set_password(
                self._SERVICE,
                self._USERNAME,
                json.dumps(data, indent=2),
            )
        except Exception as exc:
            logger.error("Failed to write credentials to keyring: %s", exc)
            raise

    def save(self, creds: DeviceCredentials) -> None:
        """Merge and persist credentials in Windows Credential Manager."""
        data = self._read()
        data.update(creds)
        self._write(data)

    def get_api_key(self) -> Optional[str]:
        """Return the stored API key or None."""
        return self._read().get("api_key")

    def get_device_id(self) -> Optional[str]:
        """Return the stored device ID or None."""
        return self._read().get("device_id")

    def get_metadata(self, key: str) -> Optional[str]:
        """Return a metadata field by key or None."""
        return self._read().get(key)

    def clear(self) -> None:
        """Remove all stored credentials from Windows Credential Manager."""
        try:
            self._kr.delete_password(self._SERVICE, self._USERNAME)
        except self._kr.errors.PasswordDeleteError:
            pass
        except Exception as exc:
            logger.warning("Failed to clear credentials from keyring: %s", exc)

    def is_provisioned(self) -> bool:
        """Return True if the keyring contains a device_id."""
        return self._read().get("device_id") is not None


# ---------------------------------------------------------------------------
# Windows file storage (JSON file under ``%PROGRAMDATA%``)
# ---------------------------------------------------------------------------


def _get_windows_credential_dir() -> Path:
    r"""Return the Windows directory for credential storage.

    Uses ``%PROGRAMDATA%\Homepot`` (all-users) or ``%APPDATA%\Homepot``
    (per-user) as a fallback.
    """
    program_data = os.environ.get("PROGRAMDATA")
    if program_data:
        path = Path(program_data) / "Homepot"
        try:
            path.mkdir(parents=True, exist_ok=True)
            return path
        except (OSError, PermissionError):
            pass
    app_data = os.environ.get("APPDATA")
    if app_data:
        path = Path(app_data) / "Homepot"
        try:
            path.mkdir(parents=True, exist_ok=True)
            return path
        except (OSError, PermissionError):
            pass
    from tempfile import gettempdir

    tmp_path = Path(gettempdir()) / "homepot"
    tmp_path.mkdir(parents=True, exist_ok=True)
    return tmp_path


class WindowsFileStorage(CredentialStorage):
    r"""Stores credentials in a JSON file under ``%PROGRAMDATA%``.

    Default location: ``%PROGRAMDATA%\Homepot\credentials``
    Fallback: ``%APPDATA%\Homepot\credentials``
    """

    def __init__(self, file_path: Optional[Path] = None) -> None:
        """Initialize Windows file storage at the given path."""
        self._file_path = file_path or _get_windows_credential_dir() / "credentials"

    def _read(self) -> Dict[str, str]:
        """Read credentials from the JSON file. Returns empty dict on error."""
        if not self._file_path.exists():
            return {}
        try:
            raw = self._file_path.read_text("utf-8")
            return dict(json.loads(raw))
        except (json.JSONDecodeError, OSError) as exc:
            logger.warning("Failed to read credentials file: %s", exc)
            return {}

    def _write(self, data: Dict[str, str]) -> None:
        """Write credentials to the JSON file."""
        self._file_path.parent.mkdir(parents=True, exist_ok=True)
        self._file_path.write_text(
            json.dumps(data, indent=2),
            encoding="utf-8",
        )

    def save(self, creds: DeviceCredentials) -> None:
        """Save credentials to the JSON file."""
        data = self._read()
        data.update(creds)
        self._write(data)

    def get_api_key(self) -> Optional[str]:
        """Return the stored API key or None."""
        return self._read().get("api_key")

    def get_device_id(self) -> Optional[str]:
        """Return the stored device ID or None."""
        return self._read().get("device_id")

    def get_metadata(self, key: str) -> Optional[str]:
        """Return a metadata field by key or None."""
        return self._read().get(key)

    def clear(self) -> None:
        """Remove the credentials file."""
        if self._file_path.exists():
            self._file_path.unlink()

    def is_provisioned(self) -> bool:
        """Return True if the file contains a device_id."""
        return self._file_path.exists() and "device_id" in self._read()


# ---------------------------------------------------------------------------
# Platform-aware factory
# ---------------------------------------------------------------------------


def create_credential_storage(
    storage_path: Optional[Path] = None,
) -> CredentialStorage:
    """Return the most appropriate ``CredentialStorage`` for the current platform.

    Priority:
    **Windows:**
    1. **WindowsCredManager** — Windows Credential Manager via ``keyring``
    2. **WindowsFileStorage** — JSON file under ``%PROGRAMDATA%``
    3. **SimulationStorage** — in-memory fallback (dev / testing)

    **Linux:**
    1. **KeyringCredentialStorage** — OS keyring via the ``keyring`` library
    2. **LinuxFileStorage** — file on disk with ``0o600`` permissions
    3. **SimulationStorage** — in-memory fallback (dev / testing)
    """
    if sys.platform == "win32":
        try:
            from importlib import import_module

            import_module("keyring")
            return WindowsCredManager()
        except (ImportError, Exception) as exc:
            logger.debug(
                "Windows Credential Manager unavailable (%s); falling back to "
                "WindowsFileStorage",
                exc,
            )
        return WindowsFileStorage(file_path=storage_path)

    if os.name == "posix" and sys.platform != "darwin":
        try:
            from importlib import import_module

            import_module("keyring")
            return KeyringCredentialStorage()
        except (ImportError, Exception) as exc:
            logger.debug(
                "Keyring unavailable (%s); falling back to LinuxFileStorage", exc
            )
        return LinuxFileStorage(file_path=storage_path)

    logger.info(
        "No platform-specific credential storage for %s; using SimulationStorage",
        sys.platform,
    )
    return SimulationStorage()
